File: tries/implementTriePrefixTree.java
//https://leetcode.com/problems/implement-trie-prefix-tree/

/* Time complexity: O(n) : insert and search
Auxiliary space: O(n)*/

package tries;

import java.io.*;
import java.util.*;

class Trie {
  
  //Trie structure
  class TrieNode{
     TrieNode [] children;
     boolean isEnd;
    public TrieNode() {
        children = new TrieNode[26];
        isEnd = false;
    }
   }
  
    private final TrieNode root;
  
    public Trie() {
        root = new TrieNode();
    }
    
  //insert function in Trie
    public void insert(String word) {
        TrieNode curr = root;
        for(int indx = 0; indx<word.length(); indx++){
            char c = word.charAt(indx);
             int ind = c - 'a';
            if(curr.children[ind] == null){
                curr.children[ind] = new TrieNode();
            }
            curr = curr.children[ind];
        }
        curr.isEnd = true;
    }
    
    //search function in Trie
    public boolean searchWord(String word) {
        TrieNode curr = root;
        for(int indx = 0; indx<word.length(); indx++){
            char c = word.charAt(indx);
             int ind = c - 'a';
            if(curr.children[ind] == null){
               return false;
            }
            curr = curr.children[ind];
        }
        return curr.isEnd == true;
    }
    
    //checking prefix function in Trie
    public boolean checkstartsWith(String prefix) {
        TrieNode curr = root;
        for(int indx = 0; indx<prefix.length(); indx++){
            char c = prefix.charAt(indx);
            int ind = c - 'a';
            if(curr.children[ind] == null){
                return false;
            }
            curr = curr.children[ind];
        }
        return true;
    }
}

File: tries/implementTriePrefixTree.py
from collections import defaultdict


class TrieNode :
    def __init__(self):
        self.children = defaultdict (TrieNode)
        self.isword = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:

        node = self.root
        for ch in word:
            node = node.children[ch]
        node.isword = True
        

    def search(self, word: str) -> bool:

        node = self.root
        for ch in word:
            if ch not in node.children: return False

            node = node.children[ch]
            
        return node.isword
        

    def startsWith(self, prefix: str) -> bool:

        node = self.root
        for ch in prefix:
            if ch not in node.children : return False

            node = node.children[ch]
        
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)